### ISSUE TITLE
Not compatible with Windows Crate 0.54

### ISSUE THREAD

**Issue by tance77** [bengreenier/win_event_hook#7]
I think some of the types have moved around and this isn't working with the latest `windows` crate

**Comment by bengreenier** [comment]
Hey @tance77 - Do you have a repro by any chance? With [this commit](https://github.com/bengreenier/win_event_hook/commit/cc89bef3d3c893ff6c62b107431a978ecdd028eb) I am still able to build without issue, and running the CLI appears to be working as expected.

**Comment by bengreenier** [comment]
Not opposed to just merging that and shipping a new build, but just curious to understand what exactly you're seeing, as it sounds like that change on it's own might not make a differrence.

**Comment by tance77** [comment]
@bengreenier  Sorry I don't have a repo to share I'm sorry I can supply some code snippets of what I am using.

```rs
    let handler = move |_ev, hwnd, _, _, _, _| {};
    let windows_hook_handle = std::thread::spawn(move || {
        let _hook = win_event_hook::WinEventHook::install(config.clone(), handler.clone()); //This line fails with windows 0.54
   }
   ```
   
   Error
   ```
   error[E0631]: type mismatch in closure arguments
   --> src\windows_system_events.rs:316:75
    |
261 |     let handler = move |_ev: Event, hwnd: HWND, _: i32, _: i32, _:u32, _:u32 | {
    |                   ------------------------------------------------------------ found signature defined here
...
316 |         let _hook = win_event_hook::WinEventHook::install(config.clone(), handler.clone());
    |                     -------------------------------------                 ^^^^^^^^^^^^^^^ expected due to this
    |                     |
    |                     required by a bound introduced by this call
    |
    = note: expected closure signature `fn(win_event_hook::events::Event, windows::Windows::Win32::Foundation::HWND, _, _, _, _) -> _`
               found closure signature `fn(win_event_hook::events::Event, HWND, _, _, _, _) -> _`
    = note: required for `{closure@src\windows_system_events.rs:261:19: 261:79}` to implement `win_event_hook::EventHandler`
note: required by a bound in `WinEventHook::install`
```

**Comment by bengreenier** [comment]
I take it you have `use windows::Windows::Win32::Foundation::HWND` in that file as well?

If so, I think that makes sense - the error is indicating we're trying to pass an older version of `HWND` to a closure that's defined to take the newer version.

if you change your handler to use [WindowHandle](https://docs.rs/win_event_hook/latest/win_event_hook/handler/type.WindowHandle.html) as the type for `hwnd` does the compilation error go away? (I realize this may not be a sufficient solution, just trying to make sure I understand the issue).

**Comment by tance77** [comment]
I have set the type of hwnd to `WindowHandle`

**Updated Handler**

```rs
let handler = move |_ev, hwnd:WindowHandle, _, _, _, _| {};
```

This worked! I'm not sure why but I'm glad it did this allows me to update the windows crate :)

**Comment by tance77** [comment]
@bengreenier I take it back I think I had some cached crates or something but after running cargo clean the problem just moved down the line.

I eventually call

```
fn get_window_title(hwnd: WindowHandle) -> String {
    let title: String;
    unsafe {
        let mut v: Vec<u16> = vec![0; 255];
        let title_len = GetWindowTextW(hwnd, &mut v);
        title = String::from_utf16_lossy(&v[0..(title_len as usize)]);
    };

    title
}

```

```GetWindowTextW``` says the HWND is the wrong type.

**Comment by bengreenier** [comment]
@tance77 - gotcha - let me try to explain what i'm pretty sure the main issue is in more detail:
- `win_event_hook::handler::WindowsHandle` points at `windows(v0.51.1)::Win32::Foundation::HWND`
- in your crate you have `windows(v0.54.0)` stuff - notably, `windows(v0.54.0)::Win32::Foundation::HWND`

 `windows(v0.51.1)::Win32::Foundation::HWND` and `windows(v0.54.0)::Win32::Foundation::HWND` aren't the same type, so function signatures that explicitly mention one, can't take the other.

`win_event_hook` mentions `windows(v0.51.1)::Win32::Foundation::HWND` for the handler signature, which is what `win_event_hook::handler::WindowsHandle` points at, so that's why your previous fix worked for the handler.

However, now that you're trying to use `win_event_hook::handler::WindowsHandle` as a `windows(v0.54.0)::Win32::Foundation::HWND` (since everything in your crate uses the `windows(v0.54.0)` signatures) it won't be permitted.

One option is an update to this library, which would align those types - I can probably do that anyway, but this'll keep happening when/if `windows` bumps versions but `win_event_hook` doesn't.

The other option is to "cast" - in both cases, `HWND` is really just a wrapper around `isize`, since in native windows land, that's [all HWND is](https://learn.microsoft.com/en-us/windows/win32/winprog/windows-data-types) - even though the windows crate wrapper is changing, the actual value and underlying concept is not (cause windows can't change this definition easily/probably ever). Since `win_event_hook::handler::WindowsHandle` and `windows(v0.54.0)::Win32::Foundation::HWND`  are both just wrappers around a `pub isize` you should be able to do `HWND(your_windows_handle_var.0)` to convert between these two types.  This approach should be resilent to crate version mismatches, unless something drastic happens (like a windows OS change, or the windows crate deciding to make the underlying `isize` non `pub` in a future release).

TLDR, try:

```rs
use windows::Win32::Foundation::HWND;
use win_event_hook::handler::WindowsHandle;

let handler = move |_ev, hwnd: WindowHandle, _, _, _, _| {
  let corrected_hwnd = HWND(hwnd.0);

  // pass corrected_hwnd to windows::* stuff
};
```

or similar

**Comment by tance77** [comment]
Excellent explanation! I agree and the casting solution to an `isize` is 100% the correct solution. Thank you for your time!

```rs
let corrected_hwnd = HWND(hwnd.0 as isize);
```

**Comment by tance77** [comment]
@bengreenier Windows Crate Changed again to 0.58 and the type has changed again. 

I could use some help again if you could spare a moment. I don't think it's a problem with the library and 

`let corrected_hwnd = HWND(hwnd.0 as isize);`

This no longer works. 

Looks like they changed it to a `*mut c_void` 
https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/Foundation/struct.HWND.html

I'm still learning rust and windows so I'm like way over my head here.

**Comment by bengreenier** [comment]
Hey @tance77 - think pinning makes sense for your use case, but for `*mut c_void` you can try something like https://stackoverflow.com/a/74983940 - I'll get around to bumping the version, but am traveling and don't have access to windows for another couple weeks, and I comically can't validate changes here on my Mac.

**Comment by tance77** [comment]
@bengreenier Thank you I will take a look at that stack overflow issue. Enjoy your travels!
